sdk/nodejs/runtime/state.ts

```diff
@@ -0,0 +1,150 @@
+    dryRun?: boolean; // whether we are performing a preview (true) or a real deployment (false).
+    testModeEnabled?: boolean; // true if we're in testing mode (allows execution without the CLI).
+    queryMode?: boolean; // true if we're in query mode (does not allow resource registration).
+    legacyApply?: boolean; // true if we will resolve missing outputs to inputs during preview.
+    cacheDynamicProviders?: boolean; // true if we will cache serialized dynamic providers on the program side.
+    organization?: string; // the name of the current organization (if available).
+    maximumProcessListeners: number; // the number of process listeners which can be registered before writing a warning.
+    /**
+     * Directory containing the send/receive files for making synchronous invokes to the engine.
+     */
+    syncDir?: string;
+}
+
+export interface Store {
+    settings: {
+        options: WriteableOptions;
+        monitor?: any | undefined;
+        engine?: any;
+        rpcDone: Promise<any>;
+        featureSupport: Record<string, boolean>;
+    };
+    config: Record<string, string>;
+    stackResource?: Stack;
+    leakCandidates: Set<Promise<any>>;
+}
+
+export class LocalStore implements Store {
+    settings = {
+        options: {
+            organization: process.env[nodeEnvKeys.organization],
+            project: process.env[nodeEnvKeys.project] || "project",
+            stack: process.env[nodeEnvKeys.stack] || "stack",
+            dryRun: process.env[nodeEnvKeys.dryRun] === "true",
+            queryMode: process.env[nodeEnvKeys.queryMode] === "true",
+            monitorAddr: process.env[nodeEnvKeys.monitorAddr],
+            engineAddr: process.env[nodeEnvKeys.engineAddr],
+            syncDir: process.env[nodeEnvKeys.syncDir],
+            cacheDynamicProviders: process.env[nodeEnvKeys.cacheDynamicProviders] !== "false",
+            legacyApply: process.env[pulumiEnvKeys.legacyApply] === "true",
+            maximumProcessListeners: 30,
+        },
+        rpcDone: Promise.resolve(),
+        featureSupport: {},
+    };
+    config = {
+        [config.configEnvKey]: process.env[config.configEnvKey] || "",
+        [config.configSecretKeysEnvKey]: process.env[config.configSecretKeysEnvKey] || "",
+    };
+    stackResource = undefined;
+
+    /**
+     * leakCandidates tracks the list of potential leak candidates.
+    */
+    leakCandidates = new Set<Promise<any>>();
+}
+
+// Get the root stack resource for the current stack deployment
+export function getStackResource(): Stack | undefined {
+    const { stackResource } = getStore();
+    return stackResource;
+}
+
+export const setStackResource = (newStackResource?: Stack) => {
+    const localStore = getStore();
+    localStore.stackResource = newStackResource;
+};
+
+declare global {
+    /* eslint-disable no-var */
+    var globalStore: Store;
+}
+
+export const getLocalStore = () => {
+    return asyncLocalStorage.getStore() as Store | undefined;
```

Comment:
If you initialize `asyncLocalStorage` as `AsyncLocalStorage<Store>()` with the type, then you shouldn't need the type assertions.

```suggestion
    return asyncLocalStorage.getStore();
```